opal/mca/memory/patcher/memory_patcher_component.c

```diff
@@ -404,6 +407,68 @@ static size_t memory_patcher_get_shm_seg_size (const void *shmaddr)
     return seg_size;
 }
 
+static size_t get_shm_size(int shmid)
+{
+    struct shmid_ds ds;
+    int ret;
+
+    ret = shmctl(shmid, IPC_STAT, &ds);
+    if (ret < 0) {
+        return 0;
+    }
+
+    return ds.shm_segsz;
+}
+#endif
+
+#if HAS_SHMAT && defined(__linux__)
+static void *(*original_shmat)(int shmid, const void *shmaddr, int shmflg);
+
+static void *_intercept_shmat(int shmid, const void *shmaddr, int shmflg)
+{
+    void *result = 0;
+
+    size_t size = get_shm_size(shmid);
+
+    if ((shmflg & SHM_REMAP) && (shmaddr != NULL)) {
+// I don't really know what REMAP combined with SHM_RND does, so I'll just
+// guess it remaps all the way down to the lower attach_addr, and all the
+// way up to the original shmaddr+size
+        uintptr_t attach_addr = (uintptr_t)shmaddr;
+
+        if (shmflg & SHM_RND) {
+            attach_addr -= ((uintptr_t)shmaddr) % SHMLBA;
+            size += ((uintptr_t)shmaddr) % SHMLBA;
```

Comment:
IMHO it should be something like this (i will probably have to fix UCX as well):
```
            attach_addr -= ((uintptr_t)shmaddr) % SHMLBA;
            size        += ((uintptr_t)shmaddr) % SHMLBA;
            size        += (SHMLBA - (size % SHMLBA)) % SHMLBA;  /* Round size up */
```

Reply:
actually, thinking again, we can't really change the size. we got the size from shmctl() so it's the actual segment size and should be aligned to SHMLBA So we probably should not change 'size' at all, just 'attach_addr' - since the segment attach point just moves by %SHMLBA, but the segment size cannot really change.

Reply:
I'm leaning toward agreeing. I agree the new mapping will have the original 'size' as its size, and it'll only occupy the range (attach_addr ..+original_size).

And when I read the manpage for REMAP I'm sure they intended to write that it replaces any mapping that existed in the address range where the new mapping goes. But they wrote more explicitly that it replaces any mapping in (addr ..+size) eg where the new mapping would go w/o RND in play.

So being logical I'd say with RND any mapping that's in (attach_addr ..+original_size) gets replaced.

I'm still maybe 5% concerned that a pedantic reading of the manpage would say that they specified the exact range that gets replaced, and the above skips a little bit on the right hand side of that range.

Is there any downside other than an unnecessary cache invalidation to my current code that uses the lower of the low ends and the higher of the high ends of the two interpretations of the manpage? My code would be telling you a certain virtual address had been invalidated when it really hadn't.

Reply:
ok, let's be pedantic